File: test/helpers/RemoteServerHelper.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package helpers;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import jakarta.servlet.MultipartConfigElement;
import jakarta.servlet.Servlet;
import jakarta.servlet.ServletOutputStream;
import jakarta.servlet.http.HttpServletResponse;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.Writer;
import java.nio.file.Files;
import java.util.List;
import java.util.Map;
import org.apache.commons.io.IOUtils;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.servlet.ServletHandler;
import play.libs.Json;

public class RemoteServerHelper {

    public static void writeResponseFromFile(HttpServletResponse response, String filePath) {
        response.setContentType("application/json");
        response.setStatus(HttpServletResponse.SC_OK);
        try (
            FileInputStream fis = new FileInputStream(new File(filePath));
            ServletOutputStream sos = response.getOutputStream()
        ) {
            IOUtils.copy(fis, sos);
            sos.flush();
        } catch (IOException e) {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
        }
    }

    public static void writeJsonResponse(HttpServletResponse response, JsonNode node, int status) {
        response.setContentType("application/json");
        response.setStatus(status);
        try {
            response.getWriter().write(node.toString());
        } catch (IOException e) {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
        }
    }

    public static void writeEmptyJsonResponse(HttpServletResponse response) {
        response.setContentType("application/json");
        response.setStatus(HttpServletResponse.SC_OK);
        try (Writer writer = response.getWriter()) {
            ObjectNode data = Json.newObject();
            data.set("data", new ObjectMapper().createArrayNode());
            writer.write(data.toString());
        } catch (IOException e) {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
        }
    }

    public static Server createAndStartServer(int port, Map<Class<? extends Servlet>, List<String>> handlers)
        throws Exception {
        return createAndStartServer(port, handlers, false);
    }

    public static Server createAndStartServer(
        int port,
        Map<Class<? extends Servlet>, List<String>> handlers,
        boolean setMultipart
    ) throws Exception {
        Server server = new Server(port);
        server.setStopAtShutdown(true);
        ServletHandler sh = new ServletHandler();
        handlers.forEach((k, v) -> v.forEach(s -> sh.addServletWithMapping(k, s)));
        if (setMultipart) {
            String path = Files.createTempDirectory("test_upload").toString();
            handlers.forEach((k, v) ->
                v.forEach(s ->
                    sh
                        .getServlet(sh.getServletMapping(s).getServletName())
                        .getRegistration()
                        .setMultipartConfig(new MultipartConfigElement(path))
                )
            );
        }
        server.setHandler(sh);
        server.start();
        return server;
    }

    public static void shutdownServer(Server server) throws Exception {
        server.stop();
    }
}

File: test/helpers/RemoteServerHelper.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package helpers

import com.fasterxml.jackson.databind.JsonNode
import jakarta.servlet.http.HttpServletResponse
import jakarta.servlet.{MultipartConfigElement, Servlet}
import org.apache.commons.io.IOUtils
import org.eclipse.jetty.ee10.servlet.{ServletContextHandler, ServletHolder}
import org.eclipse.jetty.server.{Connector, Server, ServerConnector}
import play.api.libs.json.{JsValue, Json}

import java.io.{File, FileInputStream, IOException}
import java.nio.file.Files
import scala.util.Using

object RemoteServerHelper:
  sealed trait ServletDef:
    def bind(path: String, sch: ServletContextHandler): Unit

  object ServletDef:
    final case class FromClass(clazz: Class[? <: Servlet]) extends ServletDef:
      def bind(path: String, sch: ServletContextHandler): Unit =
        sch.getServletHandler.addServletWithMapping(clazz, path)

    final case class FromInstance(servlet: Servlet) extends ServletDef:
      def bind(path: String, sch: ServletContextHandler): Unit =
        sch.addServlet(new ServletHolder(servlet), path)

  def writeResponseFromFile(response: HttpServletResponse, filePath: String): Unit =
    response.setContentType("application/json")
    response.setStatus(HttpServletResponse.SC_OK)
    try
      Using
        .Manager { use =>
          val fis = use(new FileInputStream(new File(filePath)))
          val sos = use(response.getOutputStream)
          IOUtils.copy(fis, sos)
          sos.flush()
        }
        .recover { case _: IOException =>
          response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
        }
    catch case _: IOException => response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)

  def writeJsonResponse(response: HttpServletResponse, node: JsonNode, status: Int): Unit =
    response.setContentType("application/json")
    response.setStatus(status)
    try response.getWriter.write(node.toString)
    catch case _: IOException => response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)

  def writeJsonResponse(response: HttpServletResponse, jsValue: JsValue, status: Int): Unit =
    response.setContentType("application/json")
    response.setStatus(status)
    try response.getWriter.write(jsValue.toString)
    catch case _: IOException => response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)

  def writeEmptyJsonResponse(response: HttpServletResponse): Unit =
    response.setContentType("application/json")
    response.setStatus(HttpServletResponse.SC_OK)
    try
      Using(response.getWriter) { writer =>
        val data = Json.obj("data" -> Json.arr())
        writer.write(data.toString)
      }.recover { case _: IOException =>
        response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
      }
    catch case _: IOException => response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)

  def createServer(port: Int, multipart: Boolean, defs: (ServletDef, List[String])*): Server =
    val server               = new Server(port)
    val connector: Connector = new ServerConnector(server)
    server.addConnector(connector)
    server.setStopAtShutdown(true)
    val sch = new ServletContextHandler()
    sch.setContextPath("/")
    defs.foreach { case (servletDef, paths) =>
      paths.foreach(path => servletDef.bind(path, sch))
    }
    if multipart then
      val tempPath = Files.createTempDirectory("test_upload").toString
      sch.getServletHandler.getServletMappings.foreach { mapping =>
        val servlet = sch.getServletHandler.getServlet(mapping.getServletName)
        servlet.getRegistration.setMultipartConfig(new MultipartConfigElement(tempPath))
      }

    server.setHandler(sch)
    server.start()
    server

  def shutdownServer(server: Server): Unit =
    try server.stop()
    catch case e: Exception => throw new RuntimeException("Failed to shutdown server", e)
